client.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package httprequest
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+
+	"gopkg.in/errgo.v1"
+)
+
+// Doer is implemented by HTTP client packages
+// to make an HTTP request. It is notably implemented
+// by http.Client and httpbakery.Client.
+//
+// When httprequest uses a Doer value for requests
+// with a non-empty body, it will use DoWithBody if
+// the value implements it (see DoerWithBody).
+// This enables httpbakery.Client to be used correctly.
+type Doer interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// DoerWithBody is implemented by HTTP clients that need
+// to be able to retry HTTP requests with a body.
+// It is notably implemented by httpbakery.Client.
+type DoerWithBody interface {
+	Doer
+	DoWithBody(req *http.Request, body io.ReadSeeker) (*http.Response, error)
+}
+
+// Client represents a client that can invoke httprequest endpoints.
+type Client struct {
+	// Do holds a function that will be used to actually
+	// make the HTTP request. If it is nil, http.DefaultClient.Do
+	// will be used instead. If the request has a non-empty body
+	// and Doer implements DoerWithBody, DoWithBody
+	// will be used instead of Do.
+	Doer Doer
+
+	// If a request returns an HTTP response that signifies an
+	// error, UnmarshalError is used to unmarshal the response into
+	// an appropriate error. See ErrorUnmarshaler for a convenient
+	// way to create an UnmarshalError function for a given type. If
+	// this is nil, DefaultErrorUnmarshaler will be used.
+	UnmarshalError func(resp *http.Response) error
+}
+
+var DefaultErrorUnmarshaler = ErrorUnmarshaler(new(RemoteError))
```

Comment:
I think that the docs on ErrorUnmarshaler should be sufficient, though I've added a doc comment anyway.